stream.go

```diff
@@ -256,29 +254,37 @@ func (s *stream) GetWriteOffset() protocol.ByteCount {
 	return s.writeOffset
 }
 
-// HasDataForWriting says if there's stream available to be dequeued for writing
-func (s *stream) HasDataForWriting() bool {
+// PopStreamFrame returns the next STREAM frame that is supposed to be sent on this stream
+// maxBytes is the maximum length this frame (including frame header) will have.
+func (s *stream) PopStreamFrame(maxBytes protocol.ByteCount) *wire.StreamFrame {
 	s.mutex.Lock()
-	hasData := s.err == nil && // nothing should be sent if an error occurred
-		(len(s.dataForWriting) > 0 || // there is data queued for sending
-			s.finishedWriting.Get() && !s.finSent.Get()) // if there is no data, but writing finished and the FIN hasn't been sent yet
-	s.mutex.Unlock()
-	return hasData
-}
+	defer s.mutex.Unlock()
+
+	if s.err != nil {
+		return nil
+	}
 
-func (s *stream) GetDataForWriting(maxBytes protocol.ByteCount) ([]byte, bool /* should send FIN */) {
-	data, shouldSendFin := s.getDataForWritingImpl(maxBytes)
-	if shouldSendFin {
+	frame := &wire.StreamFrame{
+		StreamID:       s.streamID,
+		Offset:         s.writeOffset,
+		DataLenPresent: true,
+	}
+	frameLen, _ := frame.MinLength(s.version) // can never error
+	if frameLen >= maxBytes {                 // a STREAM frame must have at least one byte of data
+		return nil
+	}
+	frame.Data, frame.FinBit = s.getDataForWriting(maxBytes - frameLen)
```

Comment:
We expect this function to be called quite often with no data for writing, so can we move this call and the checks below up above the stream frame allocation? The frame allocation in the packer itself was already visible in the profile.

Reply:
Github hid this comment from me, so here's my late reply: We need the frame to determine its length (`frameLen`), so we can't move this call. This was the same before this PR, the allocation just happened in the `streamFramer`.